ldon/src/store.rs

```diff
@@ -0,0 +1,658 @@
+    }
+  }
+
+  pub fn get_syn_map(&self, ptr: Ptr<F>) -> Result<Syn<F>, StoreError<F>> {
+    let mut assoc = vec![];
+    let mut next = ptr;
+
+    while let Expr::Cons(entry, cdr) = self.get_expr(next)? {
+      if let Expr::Cons(key, val) = self.get_expr(entry)? {
+        assoc.push((self.get_syn(key)?, self.get_syn(val)?));
+        next = cdr;
+      }
+      else {
+        return Err(StoreError::ExpectedMap(ptr));
+      }
+    }
+    // could test for correctness here
+    Ok(Syn::Map(Pos::No, assoc))
+  }
+
+  pub fn get_syn(&self, ptr: Ptr<F>) -> Result<Syn<F>, StoreError<F>> {
+    let expr = self.get_expr(ptr)?;
+    match expr {
+      Expr::ConsNil => Ok(Syn::List(Pos::No, vec![], None)),
+      Expr::SymNil => Ok(Syn::Symbol(Pos::No, vec![])),
+      Expr::StrNil => Ok(Syn::String(Pos::No, "".to_string())),
+      Expr::Num(f) => Ok(Syn::Num(Pos::No, f)),
+      Expr::Char(_) => Ok(Syn::Char(Pos::No, self.get_char(ptr)?)),
+      Expr::U64(_) => Ok(Syn::U64(Pos::No, self.get_u64(ptr)?)),
+      Expr::Cons(..) => self.get_syn_list(ptr),
+      Expr::StrCons(..) => Ok(Syn::String(Pos::No, self.get_string(ptr)?)),
+      Expr::SymCons(..) => Ok(Syn::Symbol(Pos::No, self.get_symbol(ptr)?)),
+      Expr::Keyword(sym) => Ok(Syn::Keyword(Pos::No, self.get_symbol(sym)?)),
+      Expr::Map(map) => self.get_syn_map(map),
+      Expr::Link(ctx, data) => self.get_syn_link(ctx, data),
+      _ => todo!(),
+      // Expr::Comm(F, Ptr<F>),             // secret, val
+      // Expr::Thunk(Ptr<F>, Ptr<F>),       // val, cont
+      // Expr::Fun(Ptr<F>, Ptr<F>, Ptr<F>), // arg, body, env
+    }
+  }
+}
+
+impl<F: LurkField> fmt::Display for Store<F> {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    writeln!(f, "{{")?;
+    for (k, v) in self.0.iter() {
+      match v {
+        Entry::Expr(x) => {
+          writeln!(f, "  {}: {},", k, x)?;
+        },
+        Entry::Cont(x) => {
+          writeln!(f, "  {}: {:?},", k, x)?;
+        },
+        Entry::Opaque => {
+          writeln!(f, "  {}: _,", k)?;
+        },
+      }
+    }
+    writeln!(f, "}}")?;
+    Ok(())
+  }
+}
+
+impl<F: LurkField> SerdeF<F> for Store<F> {
+  fn ser_f(&self) -> Vec<F> {
+    let mut exprs = Vec::new();
+    let mut opaqs = Vec::new();
+    for (ptr, entry) in self.0.iter() {
+      match entry {
+        Entry::Expr(x) => exprs.extend(x.ser_f().into_iter()),
+        Entry::Cont(_x) => todo!(),
+        Entry::Opaque => opaqs.extend(ptr.ser_f()),
+      }
+    }
+    let mut res = vec![(opaqs.len() as u64).into()];
+    res.extend(opaqs);
+    res.extend(exprs);
+    res
+  }
+
+  fn de_f(fs: &[F]) -> Result<Store<F>, SerdeFError<F>> {
+    let mut map: BTreeMap<Ptr<F>, Entry<F>> = BTreeMap::new();
+    if fs.is_empty() {
+      return Err(SerdeFError::UnexpectedEnd);
+    }
+    let opaqs: u64 =
+      fs[0].to_u64().ok_or_else(|| SerdeFError::ExpectedU64(fs[0]))?;
+    // TODO: This will break on 32-bit targets, but maybe we don't care.
```

Comment:
It will only break on 32-bit targets with stores with more than `u32::MAX` opaque pointers, which seems unlikely to ever be created. I will update the note to clarify this, and that when we compile to wasm32 we should look at adding some additional checks